File: src/chain-observers/filters/balance-change-filter.ts
import type { TAccountName } from "@hiveio/wax";
import { AccountClassifier } from "../classifiers";
import type { IAccountBalance } from "../classifiers/account-classifier";
import type { TRegisterEvaluationContext } from "../classifiers/collector-classifier-base";
import type { TFilterEvaluationContext } from "../factories/data-evaluation-context";
import { FilterBase } from "./filter-base";

export class BalanceChangeFilter extends FilterBase {
  private readonly includeInternalTransfers: boolean;

  public constructor(
    accounts: TAccountName[],
    includeInternalTransfers: boolean = false
  ) {
    super();

    this.includeInternalTransfers = includeInternalTransfers;

    this.accounts = new Set(accounts);
  }

  private readonly accounts: Set<TAccountName>;

  public usedContexts(): Array<TRegisterEvaluationContext> {
    const classifiers: Array<TRegisterEvaluationContext> = [];
    for(const account of this.accounts)
      classifiers.push(AccountClassifier.forOptions({
        account
      }));

    return classifiers;
  }

  private previousBalanceByAccount = new Map<TAccountName, IAccountBalance>();

  private parseInternalTransfers(accountName: TAccountName, balance: IAccountBalance): boolean {
    const previousBalance = this.previousBalanceByAccount.get(accountName);
    if (previousBalance === undefined)
      throw new Error("Cannot parse internal transfers before storing a previous balance");

    for(const asset in previousBalance)
      for(const type in previousBalance[asset])
        if (previousBalance[asset][type].amount !== balance[asset][type].amount) {
          this.previousBalanceByAccount.set(accountName, balance);

          return true;
        }


    this.previousBalanceByAccount.set(accountName, balance);

    return false;
  }

  public async match(data: TFilterEvaluationContext): Promise<boolean> {
    const { accounts } = await data.get(AccountClassifier);

    for(const accountName of this.accounts) {
      const account = accounts[accountName];

      if (account === undefined)
        return false;

      const previousBalance = this.previousBalanceByAccount.get(accountName);
      if (previousBalance === undefined) {
        this.previousBalanceByAccount.set(accountName, account.balance);

        continue;
      }

      if (this.includeInternalTransfers) {
        if (this.parseInternalTransfers(accountName, account.balance))
          return true;
        continue;
      }

      const changedHP = previousBalance.HP.total.amount !== account.balance.HP.total.amount;
      const changedHIVE = previousBalance.HIVE.total.amount !== account.balance.HIVE.total.amount;
      const changedHBD = previousBalance.HBD.total.amount !== account.balance.HBD.total.amount;

      this.previousBalanceByAccount.set(accountName, account.balance);

      if (changedHP || changedHIVE || changedHBD)
        return true;
    }

    return false;
  }
}

File: python/workerbee/chain_observers/filters/balance_change_filter.py
"""BalanceChangeFilter — stateful, tracks previous balance.

Mirrors src/chain-observers/filters/balance-change-filter.ts.

Uses AccountClassifier.for_options.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .._ordered_set import OrderedSet
from .filter_base import FilterBase

if TYPE_CHECKING:
    from ..classifiers.collector_classifier_base import TRegisterEvaluationContext
    from ..factories.data_evaluation_context import DataEvaluationContext


class BalanceChangeFilter(FilterBase):
    def __init__(
        self,
        accounts: list[str],
        include_internal_transfers: bool = False,
    ) -> None:
        super().__init__()
        self._include_internal_transfers = include_internal_transfers
        self._accounts: OrderedSet[str] = OrderedSet(accounts)
        self._previous_balance_by_account: dict[str, Any] = {}

    def used_contexts(self) -> list[TRegisterEvaluationContext]:
        from ..classifiers.account_classifier import AccountClassifier

        classifiers: list[TRegisterEvaluationContext] = []
        for account in self._accounts:
            classifiers.append(AccountClassifier.for_options({"account": account}))

        return classifiers

    def _parse_internal_transfers(self, account_name: str, balance: Any) -> bool:
        prev = self._previous_balance_by_account.get(account_name)
        if prev is None:
            raise RuntimeError("Cannot parse internal transfers before storing a previous balance")
        for asset_key in prev:
            for type_key in prev[asset_key]:
                if prev[asset_key][type_key].amount != balance[asset_key][type_key].amount:
                    self._previous_balance_by_account[account_name] = balance
                    return True

        self._previous_balance_by_account[account_name] = balance
        return False

    async def match(self, data: DataEvaluationContext) -> bool:
        from ..classifiers.account_classifier import AccountClassifier

        result = await data.get(AccountClassifier)
        accounts = result["accounts"]

        for account_name in self._accounts:
            account = accounts.get(account_name)

            if account is None:
                return False

            previous_balance = self._previous_balance_by_account.get(account_name)
            if previous_balance is None:
                self._previous_balance_by_account[account_name] = account["balance"]
                continue

            if self._include_internal_transfers:
                if self._parse_internal_transfers(account_name, account["balance"]):
                    return True
                continue

            changed_hp = previous_balance["HP"]["total"].amount != account["balance"]["HP"]["total"].amount
            changed_hive = previous_balance["HIVE"]["total"].amount != account["balance"]["HIVE"]["total"].amount
            changed_hbd = previous_balance["HBD"]["total"].amount != account["balance"]["HBD"]["total"].amount

            self._previous_balance_by_account[account_name] = account["balance"]

            if changed_hp or changed_hive or changed_hbd:
                return True

        return False
